frontend/test/metabase/scenarios/question/settings.cy.spec.js

```diff
@@ -138,6 +138,10 @@ describe("scenarios > question > settings", () => {
         cy.icon("play")
           .last()
           .click();
+
+        // Prevent performing actions while the query is being executed.
+        // Which caused some race condition and failed the test.
+        cy.findByText(/Doing science/).should("not.exist");
```

Comment:
This will do the trick and is more reliable than asserting on the "non-existence" of a UI element.
Btw, if you're wondering where does that intercept come from - [`visitQuestionAdHoc()`](https://github.com/metabase/metabase/blob/aa0878e372e2a1458ab0e2f2bfe592c45094e280/frontend/test/__support__/e2e/helpers/e2e-ad-hoc-question-helpers.js#L29) helper sets the "spy" / "listener" on `POST /api/dataset` route and we can intercept it later in the test "for free".

```suggestion

        cy.wait("@dataset");
```

Reply:
Thanks, I never knew about that intercept. Although, I wonder what do you mean by using spy is more *reliable* than asserting the "non-existence" of a UI element. As for me, I kind of felt like using spy is some kind of smell, because Cypress should represent what users see, so only the UI makes sense. Spying the APIs means we're leaking the implementation details to tests, which should be used as a last resort.

For me spy would make more sense when you want to:
- Simulate race condition
- Other complex states due to asynchronous actions

Maybe I'm missing something. Let me know what you think.

Reply:
Although Cypress has some default timeouts set in the config, **and** it comes with a promise of automatically (and intelligently) waiting, that didn't prove to be enough in our tests. Metabase has a set of complex XHRs firing basically after every little interaction. Some routes and/or actions have a predictable order of those requests, which we learned to use to our advantage. Btw, an anti-pattern in Cypress is to use the arbitrary wait (like `cy.wait(1000);`)

`POST /api/dataset` is one of the most commonly used ones. It's basically an ad-hoc query. If we wait for it, that means we've got the query results back. Sometimes (rarely) there is a slight delay from the time response comes back to the time browser renders the contents of that query. That's why you'll see `cy.findByTextEnsureVisible` helper in use sporadically.

> For me spy would make more sense when you want to:
>  - Simulate race condition
>  - Other complex states due to asynchronous actions

Ant that is exactly what's happening.
Cypress is infinitely faster than the real user UI interaction so we're running into race conditions. I agree it's muddying the waters by using the implementation details, but in this case that seems legit. The level of implementation details that I'd call unacceptable is when you assert on the state of some of the react components, for example. This is different, imo.

Wre're still asserting on what users see - just making sure results loaded first.

One more thing:
It's much easier to abstract code and create useful helper functions if we make sure that the request came back, than if we have to know what the contents of that response is for each scenario. For example `openOrdersTable` and `openProductsTable` are both waiting for `dataset` response (as is **any** visitQuestionAdHoc` helper). There is no way of knowing which string will be in the UI for each of those scenarios.

Sorry for the long response.
I hope it makes sense.

Reply:
@nemanjaglumac Thanks for the thorough writeup, I really appreciate it. You made your point, I also have to agree that some utils like `visitQuestionAdHoc` are really useful, and it'd be hard to only rely on the UI alone to achieve that.